docs/spec_draft.md

```diff
@@ -300,6 +316,148 @@ the IEEE-754 specification. For boolean element type, the behavior is same as
+     ```
+    * An examples assuming P = 2: `replica_groups`: `[[0, 2], [1, 3]]` and the
+      `groups` formed: `{ei(0, 0), ei(2, 0)}, {ei(0, 1), ei(2, 1)},  {ei(1, 0), ei(3, 0)}, and {ei(1, 1), ei(3, 1)}`.
+
+2. If `channel_handle` is not provided and `use_global_device_ids` $=$ `true`
+   * Invalid configuration.
+
+3. If `channel_handle` is provided and `use_global_device_ids` = false
+   * Number of groups formed: G.
+   * Size of each group: S * P.
+   * Interpretation of `replica_groups`: contains replica ids.
+   * The formation of `groups` can be demonstrated as follows:
+     ```python
+     if len(replica_groups) == 0:
+       replica_groups = all_partions_ids
+
+     for g in replica_groups:
+       for rid in g:
+         for pid in all_partions_ids:
+           group += ei(rid, pid)
+       groups += group
+     return groups
+     ```
+    * An example assuming P = 2: `replica_groups`: `[[0, 2], [1, 3]]` and the
+      `groups` formed: `{ei(0, 0), ei(0, 1), ei(2, 0), ei(2, 1)} and {ei(1, 0), ei(1, 1), ei(3, 0), ei(3, 1)}`.
+
+3. If `channel_handle` is provided and `use_global_device_ids` $=$ `true`
+   * Number of groups formed: G.
+   * Size of each group: S.
+   * Interpretation of `replica_groups`: contains global device ids.
+   * The formation of `groups` can be demonstrated as follows:
+     ```python
+     for g in replica_groups:
+       for global_device_id in g:
+           group += ei(global_device_id / P, global_device_id % P)
+       groups += group
+     return groups
+     ```
+    * An example assuming P = 2: `replica_groups`: `[[4, 5], [6, 7]]` and the
+      `groups` formed: `{ei(2, 0), ei(2, 1)} and {ei(3, 0), ei(3, 1)}`.
+
+For each `group` $\in$ `groups`, the `result`, to be visible by each member of
+`group`, is given by `concatenate(operands, all_gather_dim)`, where
+`operands` = { `operand` w.r.t each `ei(rid, pid)` following the same order in
+`group` }.
+
+### Inputs
+
+| Name                    | Type                                                                                                                                         |
+|-------------------------|----------------------------------------------------------------------------------------------------------------------------------------------|
+| `operand`               | tensor of any supported type                                                                                                                 |
+| `all_gather_dim`        | constant of type `si64`                                                                                                                      |
+| `replica_groups`        | 2-dimensional tensor constant of type `si64`                                                                                                  |
+| `channel_handle`        | constant of type struct { `channel-id`: `si64`, `channel-type`: enum of `CHANNEL_TYPE_INVALID`, `DEVICE_TO_DEVICE`, `DEVICE_TO_HOST`, `HOST_TO_DEVICE` } |
```

Comment:
Good point! Added your point in  https://github.com/openxla/stablehlo/issues/527 where this will be used.  I will address this as part of resolving it.

Reply:
Let us keep this suggestion open and take the suggestion from others. I wanted to omit this just to avoid clutter.

Reply:
In the proposed spec we are using channel_id <= 0 to denote that channel handle is missing. 

In the ODS channel_handle is  an Optional Attr, so we may choose not to provide it as well implying the same fact. 

To remove the confusion, on which option to pick, I have added the ticket   https://github.com/openxla/stablehlo/issues/649